README.md
# ZohoTools

Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/zoho_tools`. To experiment with that code, run `bin/console` for an interactive prompt.

TODO: Delete this and the text above, and describe your gem

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'zoho_tools'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install zoho_tools

## Usage

TODO: Write usage instructions here

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/zoho_tools.


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


changelog.md
# Changelog

## 1

Authentication, subscription list subscribe, unsubscribe, stubs

## 0.1

Basic api created


spec/spec_helper.rb
require "bundler/setup"
require "zoho_tools"
require 'webmock/rspec'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end

ZohoTools.configure do |config|
  config.client_id = '1000.id123'
  config.client_secret = 'secret123'
  config.accounts_api_url = 'https://accounts.zoho.eu'
  config.campaigns_api_url = 'https://campaigns.zoho.eu'
  config.callback_url = 'http://localhost:3000/webhooks/zohos'
end


spec/zoho_tools_spec.rb
RSpec.describe ZohoTools do
  it "has a version number" do
    expect(ZohoTools::VERSION).not_to be nil
  end
end


spec/zoho_tools/oauth_spec.rb
require 'spec_helper'

RSpec.describe ZohoTools::Oauth do
  describe '.authorize' do
    subject { described_class.authorize }
    it 'generates url to authorize server' do
      expect(subject).to include('https://accounts.zoho.eu/oauth/v2/auth',
                                 'client_id=1000.id123',
                                 'client_secret=secret123',
                                 'redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fwebhooks%2Fzohos',
                                 'access_type=offline',
                                 'response_type=code',
                                 'scope=ZohoCampaigns.campaign.ALL%2CZohoCampaigns.contact.ALL')
    end
  end

  describe '.authorization_code' do
    subject { described_class.authorization_code(code) }
    let(:code) { '123123' }
    it 'returns authorization codes' do
      stub_request(:post, 'https://accounts.zoho.eu/oauth/v2/token')
        .with(body: hash_including(client_id: '1000.id123',
                                   client_secret: 'secret123',
                                   code: '123123',
                                   grant_type: 'authorization_code',
                                   redirect_uri: 'http://localhost:3000/webhooks/zohos'))
        .to_return(status: 200,
                   body: { access_token: '1000.e4d172a421c1522aaf0d565b53dd.84913cdec11207537b1a649636b3',
                           refresh_token: '1000.e4d172a421c1522aaf0d565b53dd.84913cdec11207537b1a649636b36',
                           scope: 'ZohoCampaigns.campaign.ALL ZohoCampaigns.contact.ALL',
                           api_domain: 'https://www.zohoapis.eu',
                           token_type: 'Bearer',
                           expires_in: 3600 }.to_json)
      expect(subject.success?).to be_truthy
    end
  end

  describe '.refresh_access_token' do
    subject { described_class.refresh_access_token(refresh_token) }
    let(:refresh_token) { '123123' }
    it 'returns refreshed authorization codes' do
      stub_request(:post, 'https://accounts.zoho.eu/oauth/v2/token')
        .with(body: hash_including(client_id: '1000.id123',
                                   client_secret: 'secret123',
                                   refresh_token: '123123',
                                   grant_type: 'refresh_token'))
        .to_return(status: 200,
                   body: { access_token: '1000.e4d172a421c1522aaf0d565b53dd2c8d.84913cdec11207537b1a649636b3686b',
                           scope: 'ZohoCampaigns.campaign.ALL ZohoCampaigns.contact.ALL',
                           api_domain: 'https://www.zohoapis.eu',
                           token_type: 'Bearer',
                           expires_in: 3600 }.to_json)
      expect(subject.success?).to be_truthy
    end
  end
end


lib/zoho_tools/oauth.rb
module ZohoTools
  class Oauth
    OAUTH_URL = '/oauth/v2/auth'.freeze
    TOKEN_URL = '/oauth/v2/token'.freeze

    def self.authorize(access_type: 'offline', scope: %w[ZohoCampaigns.campaign.ALL ZohoCampaigns.contact.ALL])
      uri = URI.join(ZohoTools.config.accounts_api_url, OAUTH_URL)
      uri.query = URI.encode_www_form(self.auth_params.merge({ redirect_uri: ZohoTools.config.callback_url,
                                                               access_type: access_type,
                                                               response_type: 'code',
                                                               scope: scope.join(',') }))
      uri.to_s
    end

    def self.authorization_code(code)
      response = RestClient.post(URI.join(ZohoTools.config.accounts_api_url, TOKEN_URL).to_s,
                                 self.auth_params.merge(code: code, grant_type: 'authorization_code', redirect_uri: ZohoTools.config.callback_url),
                                 { accept: :json })
      ZohoTools::Response.new(response)
    end

    def self.refresh_access_token(refresh_token)
      response = RestClient.post(URI.join(ZohoTools.config.accounts_api_url, TOKEN_URL).to_s,
                                 self.auth_params.merge(refresh_token: refresh_token, grant_type: 'refresh_token'),
                                 { accept: :json })
      ZohoTools::Response.new(response)
    end

    def self.auth_params
      { client_id: ZohoTools.config.client_id,
        client_secret: ZohoTools.config.client_secret }
    end
  end
end


lib/zoho_tools/list_management.rb
module ZohoTools
  class ListManagement
    SUBSCRIBE_URL = '/api/v1.1/json/listsubscribe'.freeze
    UNSUBSCRIBE_URL = '/api/v1.1/json/listunsubscribe'.freeze

    def self.subscribe(access_token:, list_key:, contact_info:, topic_id: nil, source: nil)
      params = { listkey: list_key, resfmt: 'JSON', contactinfo: contact_info.to_json }
      params = params.merge(source: source) if source
      params = params.merge(topic_id: topic_id) if topic_id

      response = RestClient.post(URI.join(ZohoTools.config.campaigns_api_url, SUBSCRIBE_URL).to_s,
                                 params,
                                 { 'Authorization': "Zoho-oauthtoken #{access_token}" })
      ZohoTools::Response.new(response)

    end

    def self.unsubscribe(access_token:, list_key:, contact_info:, topic_id: nil)
      params = { listkey: list_key, resfmt: 'JSON', contactinfo: contact_info.to_json }
      params = params.merge(topic_id: topic_id) if topic_id

      response = RestClient.post(URI.join(ZohoTools.config.campaigns_api_url, UNSUBSCRIBE_URL).to_s,
                                 params,
                                 { 'Authorization': "Zoho-oauthtoken #{access_token}" })
      ZohoTools::Response.new(response)
    end
  end
end


lib/zoho_tools/mock/spec_helper.rb
require 'webmock/rspec'

module ZohoTools
  module SpecHelper
    class Interface
      include RSpec::Mocks::ExampleMethods
      include WebMock::API

      def stub_authorize_code(code, access_token:, refresh_token:, success: true)
        response_body = if success
                          { access_token: access_token,
                            refresh_token: refresh_token,
                            scope: 'ZohoCampaigns.campaign.ALL ZohoCampaigns.contact.ALL',
                            api_domain: 'https://www.zohoapis.eu',
                            token_type: 'Bearer',
                            expires_in: 3600 }
                        else
                          { error: 'Invalid code' }
                        end
        stub_request(:post, URI.join(ZohoTools.config.accounts_api_url, '/oauth/v2/token'))
          .with(body: hash_including(client_id: ZohoTools.config.client_id,
                                     client_secret: ZohoTools.config.client_secret,
                                     code: code,
                                     grant_type: 'authorization_code',
                                     redirect_uri: ZohoTools.config.callback_url))
          .to_return(status: 200, body: response_body.to_json)
      end

      def stub_refresh_access_tokens(refresh_token, access_token:, success: true)
        response_body = if success
                          { access_token: access_token,
                            scope: 'ZohoCampaigns.campaign.ALL ZohoCampaigns.contact.ALL',
                            api_domain: 'https://www.zohoapis.eu',
                            token_type: 'Bearer',
                            expires_in: 3600 }
                        else
                          { error: 'Invalid refresh token' }
                        end
        stub_request(:post, URI.join(ZohoTools.config.accounts_api_url, '/oauth/v2/token'))
          .with(body: hash_including(client_id: ZohoTools.config.client_id,
                                     client_secret: ZohoTools.config.client_secret,
                                     refresh_token: refresh_token,
                                     grant_type: 'refresh_token'))
          .to_return(status: 200, body: response_body.to_json)
      end

      def stub_list_subscribe(access_token:, list_key:, contact_info:, source:, success: true)
        response_body = if success
                          { status: 'success',
                            message: 'A confirmation email is sent to the user. User needs to confirm to successfully subscribe.' }
                        else
                          { error: 'Invalid code' }
                        end

        stub_request(:post, URI.join(ZohoTools.config.campaigns_api_url, '/api/v1.1/json/listsubscribe'))
          .with(body: hash_including(listkey: list_key,
                                     resfmt: 'JSON',
                                     contactinfo: contact_info.is_a?(RSpec::Mocks::ArgumentMatchers::SingletonMatcher) ? contact_info : contact_info.to_json,
                                     source: source),
                headers: { 'Authorization' => "Zoho-oauthtoken #{access_token}" })
          .to_return(status: 200, body: response_body.to_json)
      end

      def stub_list_unsubscribe(access_token:, list_key:, contact_info:, success: true)
        response_body = if success
                          { status: 'success', message: 'User successfully unsubscribed.' }
                        else
                          { error: 'Invalid code' }
                        end

        stub_request(:post, URI.join(ZohoTools.config.campaigns_api_url, '/api/v1.1/json/listunsubscribe'))
          .with(body: hash_including(listkey: list_key,
                                     resfmt: 'JSON',
                                     contactinfo: contact_info.to_json),
                headers: { 'Authorization' => "Zoho-oauthtoken #{access_token}" })
          .to_return(status: 200, body: response_body.to_json)
      end
    end

    def zoho_tools
      @zoho_tools ||= Interface.new
    end
  end
end
